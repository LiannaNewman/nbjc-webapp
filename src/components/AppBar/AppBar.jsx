import React, { useState } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { useHistory } from 'react-router-dom';

import { withStyles } from '@material-ui/core';

import MaterialAppBar from '@material-ui/core/AppBar';
import Drawer from '@material-ui/core/Drawer';
import Toolbar from '@material-ui/core/Toolbar';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import MenuIcon from '@material-ui/icons/Menu';

const styles = (theme) => ({
  root: {
    display: 'flex',
    height: 56,
    [theme.breakpoints.up('xs')]: {
      marginBottom: 20,
    },
    [theme.breakpoints.up('mobile')]: {
      marginBottom: 80,
    },
  },
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
  },
  selected: {
    background: 'rgba(98, 0, 238, 0.08) !important',
  },
  drawer: {
    width: 375,
    flexShrink: 0,
    position: 'static !important',
  },
  drawerPaper: {
    width: 375,
  },
});

const AppBar = ({
  isLoggedIn, selected, classes, routes,
}) => {
  const [showDrawer, setShowDrawer] = useState(false);
  const pageTitle = routes.find((item) => item.key === selected).label;
  const history = useHistory();

  const showDrawerItems = () => (
    routes.map((item) => {
      // todo: add injection security
      if (item.enforceLogin && !isLoggedIn) {
        return null;
      }
      const otherProps = {
        selected: item.key === selected,
        color: item.key === selected ? 'primary' : 'inherit',
      };
      return (
        <ListItem
          button
          key={item.key}
          selected={otherProps.selected}
          className={cx({ [classes.selected]: otherProps.selected })}
          onClick={() => {
            setShowDrawer(false);
            history.push(item.path);
          }}
        >
          <ListItemIcon>
            <item.icon color={otherProps.color} />
          </ListItemIcon>
          <ListItemText>
            <Typography color={otherProps.color} variant="subtitle2">
              {item.label}
            </Typography>
          </ListItemText>
        </ListItem>
      );
    })
  );

  return (
    <div className={classes.root}>
      <MaterialAppBar position="fixed" className={classes.appBar}>
        <Toolbar>
          <IconButton
            edge="start"
            color="inherit"
            aria-label="menu"
            onClick={() => setShowDrawer(!showDrawer)}
            data-testid="appbar-menu"
          >
            <MenuIcon />
          </IconButton>
          <Typography variant="h6" data-testid="appbar-title">
            {pageTitle}
          </Typography>
        </Toolbar>
      </MaterialAppBar>
      <Drawer
        open={showDrawer}
        onClose={() => setShowDrawer(false)}
        data-testid="appbar-drawer"
        className={classes.drawer}
        classes={{ paper: classes.drawerPaper }}
      >
        <Toolbar />
        <List>{showDrawerItems()}</List>
      </Drawer>
    </div>
  );
};

AppBar.propTypes = {
  isLoggedIn: PropTypes.bool,
  selected: PropTypes.oneOf(['home', 'addSpace', 'search', 'profile']),
  classes: PropTypes.shape({}).isRequired,
  routes: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    path: PropTypes.string.isRequired,
    content: PropTypes.func,
    key: PropTypes.string.isRequired,
    // eslint-disable-next-line react/forbid-prop-types
    icon: PropTypes.object,
    enforceLogin: PropTypes.bool.isRequired,
  })).isRequired,
};

AppBar.defaultProps = {
  selected: 'home',
  isLoggedIn: false,
};

export default withStyles(styles)(AppBar);
