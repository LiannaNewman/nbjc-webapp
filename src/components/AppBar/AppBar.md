## AppBar

```jsx
const [selected, setSelected] = React.useState('home');
const onNavigate = (item) => setSelected(item.key);

<AppBar isLoggedIn={false} selected={selected} onNavigate={onNavigate} />
```