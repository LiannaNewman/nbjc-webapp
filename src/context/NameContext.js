import React, { createContext, useState } from 'react';

export const NameContext = createContext();

const NameContextProvider = ({ children }) => {
  const [spaceTitle, setSpaceTitle] = useState('');
  const [spaceData, setSpaceData] = useState();

  return (
    <NameContext.Provider
      value={{
        spaceTitle, setSpaceTitle, spaceData, setSpaceData,
      }}
    >
      {children}
    </NameContext.Provider>
  );
};

export default NameContextProvider;
