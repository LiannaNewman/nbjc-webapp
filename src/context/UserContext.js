import React, { createContext, useState, useEffect } from 'react';
import getProfileChips from '../api/profileChips';

export const UserContext = createContext();

const UserContextProvider = ({ children }) => {
  const [user, setUser] = useState({
    userId: undefined,
    auth0Id: undefined,
    token: undefined,
  });

  const [userProfile, setUserProfile] = useState({
    name: undefined,
    username: undefined,
    pronouns: undefined,
    location: undefined,
    identities: undefined,
  });

  const [profileChips, setProfileChips] = useState();

  const [profileCreated, setProfileCreated] = useState({
    creationStatus: '',
  });

  useEffect(() => {
    async function fetchIdentities() {
      const results = await getProfileChips();
      setProfileChips(results.data);
    }
    fetchIdentities();
  }, []);

  return (
    <UserContext.Provider
      value={{
        user, setUser, userProfile, setUserProfile, profileChips, profileCreated, setProfileCreated,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export default UserContextProvider;
