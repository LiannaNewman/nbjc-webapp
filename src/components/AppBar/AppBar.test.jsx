import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

import AppBar from './AppBar';
import { UserContext } from '../../context/UserContext';

describe('AppBar styling', () => {
  it('should check that the lavender book logo exists and is visible', () => {
    const UserProfile = {
      name: 'demo',
      userProfile: { username: 'demoUser' },
    };
    render(
      <>
        <UserContext.Provider value={UserProfile}>
          <BrowserRouter>
            <AppBar />
          </BrowserRouter>
        </UserContext.Provider>
      </>,
    );
    const logo = screen.findByAltText('logo');
    expect(logo).toBeInTheDocument();
  });
  it('should check if the users account name is too large it be truncated and displayed instead', () => {
    const UserProfile = {
      name: 'demo',
      userProfile: { username: 'demoUserWithAReallyLogName' },
    };
    render(
      <>
        <UserContext.Provider value={UserProfile}>
          <BrowserRouter>
            <AppBar />
          </BrowserRouter>
        </UserContext.Provider>
      </>,
    );
    const userMenu = screen.getByText('demoUserWi....');
    expect(userMenu).toBeInTheDocument();
  });
});

describe('AppBar with a logged in user', () => {
  const UserProfile = {
    name: 'demo',
    userProfile: { username: 'demoUser' },
  };
  it('should check if the username appears and clicking it opens a dropdown menu', () => {
    render(
      <>
        <UserContext.Provider value={UserProfile}>
          <BrowserRouter>
            <AppBar />
          </BrowserRouter>
        </UserContext.Provider>
      </>,
    );
    const userMenu = screen.getByText('demoUser');
    expect(userMenu).toBeInTheDocument();
    userEvent.click(userMenu);
    const MyProfileButton = screen.getByText('My Profile');
    const LogOutButton = screen.getByText('Sign Out');
    expect(MyProfileButton).toBeInTheDocument();
    expect(LogOutButton).toBeInTheDocument();
  });
  it('clicking on the profile button in the users dropdown should redirect them to the proper page', async () => {
    render(
      <>
        <UserContext.Provider value={UserProfile}>
          <BrowserRouter>
            <AppBar />
          </BrowserRouter>
        </UserContext.Provider>
      </>,
    );
    const userMenu = screen.getByText('demoUser');
    userEvent.click(userMenu);
    const MyProfileButton = screen.getByText('My Profile');
    userEvent.click(MyProfileButton);
    await waitFor(() => expect(window.location.href).toBe('http://localhost/profile'));
  });
});

describe('AppBar without logged in user', () => {
  it('should check if a user is not logged in that the Log In button displays', () => {
    const UserProfile = {
      name: undefined,
      userProfile: { username: undefined },
    };
    render(
      <>
        <UserContext.Provider value={UserProfile}>
          <BrowserRouter>
            <AppBar />
          </BrowserRouter>
        </UserContext.Provider>
      </>,
    );
    const logInMenu = screen.getByText('Log In');
    expect(logInMenu).toBeInTheDocument();
  });
  it('should check if the log in button exists and has the correct href', async () => {
    const UserProfile = {
      name: undefined,
      userProfile: { username: undefined },
    };
    render(
      <>
        <UserContext.Provider value={UserProfile}>
          <BrowserRouter>
            <AppBar />
          </BrowserRouter>
        </UserContext.Provider>
      </>,
    );
    const logInButton = screen.getByText('Log In', { selector: 'a' });
    expect(logInButton).toBeInTheDocument();
    expect(logInButton.closest('a')).toHaveAttribute('href', '/profile');
  });
});
