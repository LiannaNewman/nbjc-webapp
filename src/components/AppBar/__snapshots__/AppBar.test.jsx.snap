// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar renders with default state 1`] = `
<DocumentFragment>
  <div
    class="AppBar-root-1"
    data-testid="app-bar"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary AppBar-appBar-2 mui-fixed MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          aria-label="menu"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
          data-testid="appbar-menu"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root MuiTypography-h6"
          data-testid="appbar-title"
        >
          Home
        </h6>
      </div>
    </header>
  </div>
</DocumentFragment>
`;
